### DISCUSSION TITLE
Very basic failure to understand

### DISCUSSION THREAD

**Discussion by dd-b** [ImageMagick/ImageMagick#4675] [Help] [answered]
I'm failing to find documentation that really teaches me how the command-line stuff works -- image lists, stacks, parens, and such.

I think I need to use that, or something like it -- I'm writing a script to cut a bunch of small images out of one single big image. (The big image is an old photographic "contact sheet" showing an entire roll of film; it has been carefully aligned already so that the spacing between frames both horizontally and vertically is constant). 

I can use *individual* commands to get what I want; this gets me the two frames specified:

> $ magick ddb_403-contact.jpg -crop 898x900+100+100 "out01.jpg"
> $ magick ddb_403-contact.jpg -crop 898x900+998+100 "out02.jpg"

But...I should get much better performance if I avoid re-loading the big image for each frame I'm going to clip out of it, right? And it takes some resources to start up magick each time. So I'm trying to figure out how to combine those two lines into _one_ line that will do both crops and save both files. And everything I try gets one sort of error or another. (My actual use-case is up to 36 frames being clipped out of one contact sheet, but if I get the case of 2 working I ought to be able to extend it from there!). 

It first looked like just using +write <filename> for each file would work; but it doesn't. I have no idea why not, the docs seem to say this should work. The errors don't make any sense to me, either.

> $magick ddb_403-contact.jpg  -crop 898x900+100+100 +write "out01.jpg" -crop 898x900+998+100 +write "out02.jpg"
> magick: geometry does not contain image `ddb_403-contact.jpg' @ warning/transform.c/CropImage/645.
> magick: MissingArgument `+write' at CLI arg 8 @ fatal/magick-cli.c/ProcessCommandOptions/447.

More complicated structures using parentheses also fail, sometimes with similar errors sometimes not.

This has gotta be really simple, I'm just failing to understand how to combine operations on the cli. Help!

(I'm also confused about utility names -- 'convert' vs. 'magick'. I'm using 7.0.11; does that mean I should be using the 'magick' utility name for this, and 'convert' exists only for backwards compatibility? )

I'm running this on FreeBSD 11.3.

**Comment by fmw42** [comment]
Be sure to add +repage after each crop.

**Comment by snibgo** [comment, marked as answer]
> I'm failing to find documentation that really teaches me how the command-line stuff works -- image lists, stacks, parens, and such.

I suggest https://legacy.imagemagick.org/Usage/basics/ . That was written for v6, and we are now at v7, but the basics haven't changed much.

Operations are carried out in the order you give them. Do this, and then do this, and then do this. Sounds reasonable? So your command:
```
magick ddb_403-contact.jpg -crop 898x900+100+100 +write "out01.jpg" -crop 898x900+998+100 +write "out02.jpg"
```
... reads an image from a file. Then it crops that imge, so the image in memory is now 898x900 pixels. Then it writes that image (the result of the crop, which is 898x900 pixels) to a file. Then it tries to crop that 898x900 image. But the offset +998+100 is outside the current canvas.

What you really want to do is crop the input image and write that to a file, and do nothing else with that crop. Then take a different crop _of the input image_, and write that to a file, and do nothing else with that crop. I would do that like this (bash syntax):
```
magick \
  ddb_403-contact.jpg -write mpr:INP +delete \
  mpr:INP -crop 898x900+100+100 +repage +write "out01.jpg" +delete ^
  mpr:INP -crop 898x900+998+100 +repage +write "out02.jpg" +delete ^
  NULL:
```
I have stored a copy of the input in mpr:INP, which is in memory. So each crop is of that input image.

Repeat the line that starts "mpr:INP" as many times as you want.

**Reply by dd-b** [reply]
I saw the mpr: stuff in the documentation, that looked hopeful. I even know what the first line does there! Ookay, your second line is very like something I did that didn't quite work. The explicit +delete is an alternative to using parens I think? Experiment seems to show that those "^" should be "\\" like the first 2 lines? (I suspect you created this example on Windows :-) ).  With that fix it works fine.

When I found (after much hard work; no way to search for it) the doc on +write vs. -write it seemed to say that +write got rid of the copy it just wrote, but your script seems to think it needs to explicitly delete that copy (and it works, so you must be right). I know generally the '+' prefix resets things to a default state, but it's not always clear what that mans for a given operator.

And the NULL: gets rid of everything left at the end which avoids a warning? (I think it warns you if you create something and leave it sitting there at the end, since that is usually a mistake; when it isn't, using NULL: avoids the warning?)

I may look more closely at tiled cropping also, depending on whether I can control the output filename with sufficient precision, but this lets me get my script working the way I envisioned it, so I have marked this as an answer. Thank you!

**Comment by fmw42** [comment]
If all your images are in a uniform array, then you can do tiled cropping. See https://legacy.imagemagick.org/Usage/crop/#crop_tile

If on a Unix-like system, then you could try one of my scripts:  multicrop or multicrop2 at my web site http://www.fmwconcepts.com/imagemagick

**Reply by dd-b** [reply]
Tiled cropping was something I looked at, it might be a way to move the computation of the array size out of my script and into Image Magick. How much control do I have on output file naming with tiled cropping? Is the sequence number always at the end of the name (well, before the extension)? I could always invoke a separate rename step, but the purpose of this script for me is to get a cleaner workflow. The output file names need to be of the pattern "ddb_403-\<seq\>-thumb.jpg" to fit within my system.

I found your scripts page earlier. Multicrop might actually work if I disabled the rotation part, but it's doing a much more complicated process than I need. I was initially hoping I could get something like _that_ to work, so that I didn't need to manually go through the process of aligning each strip of negatives, but based on some reading and some experiments there isn't nearly enough distinction between the black paper background and the negative edges, or even the frame edges (lots of my photos are shot at night in dark rooms, music sessions and such, so the edges of the real frame are pretty dark). (Dunno if you've seen "contact sheets"; in the old darkroom days you put your negatives on a sheet of paper, put a sheet of glass over it, and expose the whole sandwich. Then process the paper, and you've got a positive image of each negative on the roll, all on one sheet, and all exposed the same so comparisons between them are useful. The bare paper is solid black, the film edges are just a hair short of solid black.)

Must also remember to review your scripts more as a way to learn how to use this thing!  That looks to be a great resource. My environment for this precise operation is FreeBSD, I use Linux a lot too, and I've got Windows for the Adobe products but also have MSYS2 on the windows platform, so I can actually use your scripts, not just study them.

**Reply by dd-b** [reply]
%d lets me place the sequence number but doesn't seem to let me specify the length. Since there are nearly always more than 10, I need 2-digit numbers with leading zeros (else the filenames sort wrong). I can't find any general documentation on the % escapes or exactly where they might be valid (still learning my way around the docs!).

**Reply by fmw42** [reply]
%02d should do that if you have less than 100 output files

**Reply by dd-b** [reply]
Yeah, double digits but not triple (rolls of film for 35mm still cameras were most commonly 20 and 36 exposures, though 12 and 24 also sometimes existed, and there was a weird outlier at 72 that I never used). 36 exposures divided into 6 strips of 6 negatives just barely fits on an 8x10 sheet of paper for a contact print (8x10, not letter-size, being the standard darkroom photo paper).

**Comment by snibgo** [comment]
>  (I suspect you created this example on Windows :-) ).

Oops, yes, guilty as charged, sorry. I use Windows BAT more than I use bash.

> +delete is an alternative to using parens I think?

Well, kind of. "+delete" removes the last image from the list. At each of those, there is only one image in the list, so "+delete" empties the list.

Each line...
```
mpr:INP -crop 898x900+100+100 +repage +write "out01.jpg" +delete
```
... starts with an empty image list. It reads an image, crops it, writes that to a file, and removes the image from the list. So we can have many more similar lines without the list continually growing. If we didn't have "+delete", then the next line would already have an image in the list, so "mpr:INP" would read a second image, "-crop" would crop both images, and "+write" would write both images. Then the next line would have three images, and so on.

There are usually alternative commands to do the same thing. For example:
```
magick \
  ddb_403-contact.jpg \
  ( +clone -crop 898x900+100+100 +repage +write "out01.jpg" +delete ) \
  ( +clone -crop 898x900+998+100 +repage +write "out02.jpg" +delete ) \
  NULL:
```
Each open parenthesis pushes the current image list on a stack and starts a new empty list. The "+clone" copies the last image from the outer list to the inner list. Each close parens closes the new list, pops the outer list from the stack, and copies any images from the inner list to the outer list. (In this command, the inner list is empty at the close parens.)

Internally, "-write" used to do this:
1. Modify the image(s) in the list to make them suitable for the output format, eg reducing bit depth.
2. Write the image(s) to file(s).

So subsequent operations would use the modified images. By contrast, "+write" did this:
1. Make temporary clones of the image(s) in the list.
2. Modify those clones to make them suitable for the output format, eg reducing bit depth.
3. Write the clone(s) to file(s).
4. Delete the temporary clones.

So images in the list were not modified.

I suspect the behaviour has changed, and "-write" now does the same as "+write", but I'm not sure.

The "+" prefix restores a _setting_ to the default state. "write" is an _operation_, not a setting. Many operations have "-" and "+" versions that do slightly different things.

> And the NULL: gets rid of everything left at the end which avoids a warning?

Yes.

As fmw42 suggests, if your contact sheet is a regular grid, that is a simpler solution, something like this.
```
magick \
  ddb_403-contact.jpg \
  -crop 898x900 +repage \
  -crop WxH+X+Y +repage \
  outfile.ext
```

**Reply by dd-b** [reply]
And he just pointed out how to format the name the way I need, too. 

Okay, let's see if I understand that. Your simpler solution is creating a  stack of images using crop, but not doing anything with each one, so when it comes to the output file name the stack is waiting to be written?  It's useful to have all the parameters for each crop set individually -- I'm finding that the image spacing is not absolutely consistent across cameras (I shouldn't be surprised; the film transports were pretty simple mechanical devices, not super-accurate), so I'm thinking of cropping each image a bit oversize, but still moving it in standard units (so, possibly, the crop areas would overlap -- not a problem in that above sequence as I understand it).

Plus, I kind of hate using a tool by rote when I don't understand how the recipes I've memorized really work :-) .

**Comment by snibgo** [comment]
On contact sheets: I have a shelf full of ring binders bulging with contact sheets. I haven't digitized them, and probably won't. Life is too short.

If I did, this would be my approach. Each sheet has 6 strips each with 6 negatives. The strips are mostly roughly parallel to each other, but sometimes with an extra strip of a couple of negatives at a right-angle, and there is often overlap between strips.

1. Create a data file for each sheet. This would contain one line per negative strip. Each line would contain the coordinates of the top-left corner and bottom-right corner of the strip.

2. A script can then extract each strip, and rotate it to make it horizontal. The coordinates indicate whether the strip has 6 negs or some smaller number, so it can chop it into individual negs.

**Reply by dd-b** [reply]
While I'm not _surprised_ to find photographers old enough to have that experience here, I wasn't assuming _everyone_ here was. 

Life is too short for so many thing, but I'm putting most of my work now into historical photos of science-fiction fandom, so these old contact sheets are pretty important for that. It's a fast way to get thumbnail scans that we can then use to figure out which photos we care enough about to do real scans of the negatives, and which it's just handy to be able to quickly check are of no importance. Plus -- when I get up to the mid 1980s, in my own collection there are a few years where I've just got the negatives in envelopes, and I can do contact-sheet scans directly (I've got a scanner with 8x10 transparency adapter). 

I did shoot one big project and self-publish a photo book back in 2020, so I'm not _entirely_ living in the past photographically.

**Reply by dd-b** [reply]
Interesting script idea. That would, I think, be less work than what I've been doing (importing the scanned full-page into a Photoshop template with marks for where the strips should go, and manually moving each strip to where it goes, including straightening). Although I would need a tool to let me copy the coordinates when I click on the corner of the strip, which I don't think I have now. Could be written in Perl I guess using Qt or something, but I'm not very familiar with the tools for GUI work in Perl (or, really, anything else; I'm a back-end programmer), so writing that would be another learning curve. That tool could immediately do the crops, too, rather than writing a script.

**Comment by snibgo** [comment]
> Your simpler solution is creating a stack of images using crop, but not doing anything with each one, so when it comes to the output file name the stack is waiting to be written?

I can't be sure in advance how I would do job, because problems often only show themselves after the work has started.

My approach is: from each contact sheet, first create one image per negative strip, then chop each strip into frames. Depending on the purpose, chopping strips into frames may not be important. Remember that one benefit of contact sheets is that we see each image in context.

> ... image spacing is not absolutely consistent across cameras...

"Full-frame" 35mm cameras have 8 sprocket-holes per frame, so that defines the average spacing from one frame to the next, but mechanical slippage creates variation. Some cameras are so sloppy that some frames actually overlap. The actual frame size, nominal 36x24mm, varies between cameras, eg screw-thread Leicas have larger frames than Nikon F.

Some cameras are not full-frame, eg Pen-F half-frame, stereo, Tessina, and so on.

**Reply by dd-b** [reply]
Yeah, no plan survives contact with the enemy!

Since I'm no longer doing darkroom printing, comparing exposures isn't important. I can see the photos in the context of the roll after chopping them up in the grid view of Lightroom. And chopping them up means finding photos tagged to individual people works usefully.  So, while it's work, it's a win, it may be the easiest way to get usable thumbnails onto the web site.

Yeah, small slop, plus small variation in frame sizes (9 sprocket holes, though, from start to start of a frame). A bunch of these rolls were actually shot on a Leica M3, which has the same net spacing but narrower gaps between frames (slightly larger frames). 

I've got other formats entirely, medium format up to 6x9, a very little 4x5, and even some Pocket Instamatic 110. No half-frame, no Minox, no Minolta 16mm, in the actual collection. I do have arguments to my script to give frame spacing and strip width in scanned pixels, but the defaults are set for 35mm.

Been looking at WxWidgets and Perl integration -- and finding I can't figure out even how to install it (and there don't seem to be packages). Back to that "learning curve" idea. Installation might be easier for Python, looks like, so I could write that one utility in Python (which I'm not expert in, but have done stuff over the years). I dunno. 

The actual GUI utility would I think be quicker to use, and also much more generally useful (since it would actually work with any random collection of stuff together on a page; using the human to do the hard visual processing). But it would be day or, realistically, weeks, of delaying other stuff, and in those weeks I could do rather a lot of contact sheets my less-good way. I've done chopping up contact sheets _completely_ manually before, and the results have been very useful.

From 1981, Steven Brust and Emma Bull tuning up for a music session:

![skzb-emma-thumb](https://user-images.githubusercontent.com/4041966/148613197-d58328f3-2f08-4638-af10-b8ea860dc8e7.jpg)
